pkg/apis/ceph.rook.io/v1/labels.go

```diff
@@ -16,41 +16,73 @@ limitations under the License.
 
 package v1
 
-import (
-	rook "github.com/rook/rook/pkg/apis/rook.io/v1"
-)
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+func (a LabelsSpec) All() Labels {
+	return a[KeyAll]
+}
+
+// ApplyToObjectMeta adds labels to object meta unless the keys are already defined.
+func (a Labels) ApplyToObjectMeta(t *metav1.ObjectMeta) {
+	if t.Labels == nil {
+		t.Labels = map[string]string{}
+	}
+	for k, v := range a {
+		if _, ok := t.Labels[k]; !ok {
+			t.Labels[k] = v
+		}
+	}
+}
+
+// Merge returns a Labels which results from merging the attributes of the
+// original Labels with the attributes of the supplied one. The supplied
+// Labels attributes will override the original ones if defined.
+func (a Labels) Merge(with Labels) Labels {
+	ret := Labels{}
+	for k, v := range a {
+		if _, ok := ret[k]; !ok {
+			ret[k] = v
+		}
+	}
+	for k, v := range with {
+		if _, ok := ret[k]; !ok {
+			ret[k] = v
+		}
+	}
+	return ret
+}
```

Comment:
`Labels` is such a low-level type. If we have shared functionality that goes along with it like this, I don't understand why we don't keep it in the `rook.io/v1` package for other backends to reuse if they choose.

I definitely agree that having aggregate types in `rook.io/v1` is sometimes a challenge, but I think there is real use to have the low-level types like this there.

Reply:
Ok after a bit more experimentation, I've moved the annotations and labels to the `rook.io` package. The version subdirectories such as `v1` just don't seem like the right place for these basic types. I could consider a `rook.io/core`, `rook.io/common`, or similar package, but directly in `rook.io` seems just as appropriate. Thoughts?

Reply:
I think that seems pretty reasonable. 👍